lib/api/controllers/securityController.js

```diff
@@ -753,3 +904,140 @@ function formatUserSearchResult(kuzzle, formatter, raw) {
       return users;
     });
 }
+
+
+/**
+ * Creates a new User object in Kuzzle's database layer
+ *
+ * @param {Kuzzle} kuzzle
+ * @param {Request} request
+ * @param {boolean} restricted
+ * @returns {Promise<Object>}
+ */
+function createUser (kuzzle, request, restricted) {
+  let pojoUser;
+
+  assertHasBody(request);
+  assertIdStartsNotUnderscore(request);
+  assertBodyHasAttribute(request, 'content');
+  assertBodyAttributeType(request, 'content', 'object');
+
+  if (request.input.body.credentials) {
+    assertBodyAttributeType(request, 'credentials', 'object');
+  }
+
+  if (restricted) {
+    if (request.input.body.content.profileIds) {
+      throw new BadRequestError(`${getRequestRoute(request)} must not specify the body attribute "content.profileIds".`);
+    }
+  }
+  else {
+    if (!request.input.body.content.profileIds) {
+      throw new BadRequestError(`${getRequestRoute(request)} must specify a body attribute "content.profileIds".`);
+    }
+    if (!Array.isArray(request.input.body.content.profileIds)) {
+      throw new BadRequestError(`${getRequestRoute(request)} must specify the body attribute "content.profileIds" of type "array".`);
+    }
+  }
+
+  pojoUser = request.input.body.content;
+  pojoUser._id = request.input.resource._id || uuid.v4();
+
+  if (restricted) {
+    pojoUser.profileIds = kuzzle.config.security.restrictedProfileIds;
+  }
+
+  return Bluebird.coroutine(persistUserAndStrategies)(kuzzle, request, pojoUser);
+}
+
+/**
+ * @param {Kuzzle} kuzzle
+ * @param {Request} request
+ * @param {object} pojoUser
+ */
+function* persistUserAndStrategies (kuzzle, request, pojoUser) {
+  const
+    user = new User(),
+    loadedUser = yield Bluebird.resolve(kuzzle.repositories.user.load(pojoUser._id));
+  let
+    strategies = [],
+    errorStep = null,
+    creationError,
+    createdUser,
+    errors;
+
+  if (loadedUser !== null && loadedUser._id !== kuzzle.repositories.user.anonymous()._id) {
+    throw new BadRequestError(`${getRequestRoute(request)} user "${pojoUser._id}" already exist.`);
+  }
+
+  if (request.input.body.credentials && Object.keys(request.input.body.credentials).length > 0) {
+    strategies = Object.keys(request.input.body.credentials);
+
+    for (let i = 0; i < strategies.length; i++) {
+      const strategy = strategies[i];
+      let validateMethod;
+
+      if (kuzzle.pluginsManager.listStrategies().indexOf(strategies[i]) === -1) {
+        throw new BadRequestError(`${getRequestRoute(request)} strategy "${strategies[i]}" is not a known strategy.`);
+      }
+
+      validateMethod = kuzzle.pluginsManager.getStrategyMethod(strategy, 'validate');
+
+      try {
+        yield validateMethod(request, request.input.body.credentials[strategy], pojoUser._id, request.input.args.strategy, false);
+      }
+      catch (error) {
+        throw new BadRequestError(error);
+      }
+    }
+
+    for (let i = 0; i < strategies.length; i++) {
```

Comment:
Array.every as I'd like to stop as soon as possible. subsequent: how can I do to make it work with generators/coroutine